fecgov/openFEC
# Add leadership sponsor and org type to /totals/by_entity/

## Summary (required)

- Resolves #4879

Add filters to /totals/entity_type/ to support PAC/Party totals data table: 
- `sponsor_candidate_id` filter and `sponsor_candidate_ids` output field (only entity_type `pac` and `pac-party`)
- `organization_type` output and filter, `organization_type_full` output field
- `sponsor_candidate_list` output field 

### Required reviewers

- @hcaofec for database changes
- @fec-jli and/or @pkfec for application changes

## Impacted areas of the application

General components of the application that this PR will affect:

`/totals/entity_type` and PAC and party data table 

### Database changes

**V0233**

1 - Modify `ofec_totals_combined_mv` and `ofec_totals_combined_vw` to add:
    `sponsor_candidate_ids`
    `organization_type` and index
    `organization_type_full`

    Change index on `committee_designation_full` to `committee_designation`
    Change index on `committee_type_full` to `committee_type`

    Replaces V0232

2 - Modify `ofec_totals_pac_party_vw` to add:
    `sponsor_candidate_ids`
    `organization_type`
    `organization_type_full`

    Replaces V0232

3 - Modify `ofec_committee_totals_per_cycle_vw` to add:
    `organization_type`
    `organization_type_full`

    Replaces V0232

4 - Modify `ofec_totals_house_senate_mv` and `ofec_totals_house_senate_vw` to add:
    `organization_type` and index
    `organization_type_full`

    Replaces V0232

    Change index on `committee_type_full` to `committee_type`
    Change index on `committee_designation_full` to `committee_designation`

5 - Modify `ofec_totals_ie_only_mv` and `ofec_totals_ie_only_vw` to add:
    `organization_type` and index
    `organization_type_full`

    Replaces V0232

    Change index on `committee_designation_full` to `committee_designation`
    Change index on `committee_type_full` to `committee_type`

**V0234**

-- View: public.ofec_pac_sponsor_candidate_per_cycle_vw
-- This view is created for ticket #4879.
-- This view includes historical data for leadership PAC committees and their sponsors

## How to test
- Check out this branch
- Test migration file and MV refresh with `createdb cfdm_test` / `dropdb cfdm_test` / `invoke create_sample_db`
- Connect to the `dev` database
- Modify the following classes in your local `openFEC/webservices/common/models/totals.py` to point to the temporary testing views (creation script here: https://github.com/fecgov/openFEC/issues/4879#issuecomment-877329656)

`openFEC/webservices/common/models/candidates.py` 

```python
PacSponsorCandidatePerCycle(db.Model):
-    __tablename__ = "ofec_pac_sponsor_candidate_per_cycle_vw"
+    __tablename__ = "ofec_pac_sponsor_candidate_per_cycle_vw_lb"
```
 
`openFEC/webservices/common/models/totals.py`

```python
 class CommitteeTotalsPacParty(CommitteeTotals):
-    __tablename__ = 'ofec_totals_pac_party_vw'
+    __tablename__ = 'ofec_totals_pac_party_vw_lb'
 
 class CommitteeTotalsHouseSenate(CommitteeTotals):
-    __tablename__ = 'ofec_totals_house_senate_mv'
+    __tablename__ = 'ofec_totals_house_senate_mv_lb'
 
 class CommitteeTotalsIEOnly(BaseModel):
-    __tablename__ = 'ofec_totals_ie_only_mv'
+    __tablename__ = 'ofec_totals_ie_only_mv_lb'
 
 class CommitteeTotalsPerCycle(CommitteeTotals):
-    __tablename__ = 'ofec_committee_totals_per_cycle_vw'
+    __tablename__ = 'ofec_committee_totals_per_cycle_vw_lb'
 
```
- Run some test filter queries like these: https://docs.google.com/spreadsheets/d/1NIj_Fcmby10_wLIZ2wI4eOmVp47n08SN_keGZsWkDWY/edit#gid=0 (focus on pac-party entity)
- Test existing filters like `cycle` and `committee_id`

## Related PR
https://github.com/fecgov/openFEC/pull/4900
https://github.com/fecgov/openFEC/pull/4893
